### DISCUSSION TITLE
Google Gemini API

### DISCUSSION THREAD

**Discussion by kristopher-wood** [evennia/evennia#3384] [Community Contribs & Snippets]
Hi all,

I've put some time in on getting the Google Gemini API working within Evennia, initially based on the LLM NPC contrib by Griatch. Currently I'm stuck on getting it to act more like a player character. It doesn't understand the special characters in the msg output and thinks they're part of the message. It also doesn't understand that it's a character in a game. Work-In-Progress!
[geminiNPC.zip](https://github.com/evennia/evennia/files/13775235/geminiNPC.zip)

**Instructions**
Put the contents of the zip file in your game's typeclasses directory.
You'll need to add your Google Gemini API to your settings.py:

```Python
GOOGLE_API_KEY = "YoUrApIKeYHeRe"
```

I'm not entirely sure that's working yet, so you can alternatively add it to geminiClient.py line 22

```Python
DEFAULT_GOOGLE_API_KEY = "YoUrApIKeYHeRe"
```

In-game I've tested this by adding the new character as a default object...

```create/drop Gemini```

...then convert it to a GeminiNPC

```type/update Gemini=typeclasses.geminiNPC.GeminiNPC```

Currently the NPC reacts to things that happen in its room, but it's not fully working. You can see its reactions in the server log.

Before I added `at_msg_receive` the `at_talked_to` was working. I'm not sure if it is at the moment because I stopped working on it. If so you can try talk Gemini hello or whatever.

**Comment by kristopher-wood** [comment]
Sorry, it's my first time sharing code on a public repo discussion, so I'm new to how to do it properly. I'll edit the first post to add instructions.

**Comment by jamesmpls80** [comment]
Hi, I am implementing this code ona pre-alpha MUD called Sphere. I fixed some parts of the code so it correctly calls gemini, sends what is happening in the room, and outputs the responses from gemini into the terminal.  It is buggy and inconsistent, but it doesnt crash, and is highly prone to bizarre hallucinations.  It thinks it is supposed to be writing a movie/tv/play script because of the feedback from the teminal, so it starts narrating an imaginary story.  You can find it on the evennia website.  I can also update the script that works if that is of interest to the community.

**Comment by InspectorCaracal** [comment]
Funny enough, I've been doing some stuff with the LLM contrib too, lately! 😆 I've been specifically focusing on implementing support for instruct templates and AI Horde, not big-name chat completion APIs like Gemini, but I can tidy up my WIP a bit to share in here if there's interest.

I've generally found that when you're building your own prompt, if the model's response is way worse than you expect from that model in terms of following instructions and hallucinations, it means that the prompt format is the wrong one for the model or that the parameters are way off. I know the core LLM contrib was only tested on local text completion, so that might be relevant.

**Comment by jamesmpls80** [comment]
I'm back to trying this again, but with an openai API implementation.  I realized the major issue was that the instructions passed to gemini were not detailed enough, and gemini is really not a very good model for roleplay characters.  If I get the openai API to work well, I''ll share my results here.
